src/test/request_cancellation_test.cc

```diff
@@ -0,0 +1,150 @@
+    TRITONSERVER_ServerOptions* server_options = nullptr;
+    FAIL_TEST_IF_ERR(TRITONSERVER_ServerOptionsNew(&server_options));
+    FAIL_TEST_IF_ERR(TRITONSERVER_ServerOptionsSetModelRepositoryPath(
+        server_options, "./models"));
+    FAIL_TEST_IF_ERR(TRITONSERVER_ServerOptionsSetBackendDirectory(
+        server_options, "/opt/tritonserver/backends"));
+
+    FAIL_TEST_IF_ERR(TRITONSERVER_ServerNew(&server_, server_options));
+    FAIL_TEST_IF_ERR(TRITONSERVER_ServerOptionsDelete(server_options));
+  }
+
+  static void TearDownTestSuite()
+  {
+    FAIL_TEST_IF_ERR(TRITONSERVER_ServerDelete(server_));
+  }
+
+  void SetUp() override
+  {
+    ASSERT_TRUE(server_ != nullptr) << "Server has not created";
+    // Wait until the server is both live and ready.
+    size_t health_iters = 0;
+    while (true) {
+      bool live, ready;
+      FAIL_TEST_IF_ERR(TRITONSERVER_ServerIsLive(server_, &live));
+      FAIL_TEST_IF_ERR(TRITONSERVER_ServerIsReady(server_, &ready));
+      if (live && ready) {
+        break;
+      }
+
+      if (++health_iters >= 10) {
+        FAIL() << "failed to find healthy inference server";
+      }
+
+      std::this_thread::sleep_for(std::chrono::milliseconds(500));
+    }
+  }
+
+
+  static TRITONSERVER_Server* server_;
+};
+
+TRITONSERVER_Server* RequestCancellationTest::server_ = nullptr;
+
+
+TEST_F(RequestCancellationTest, Cancellation)
+{
+  TRITONSERVER_InferenceRequest* request;
+  FAIL_TEST_IF_ERR(
+      TRITONSERVER_InferenceRequestNew(&request, server_, "model", 1));
+  FAIL_TEST_IF_ERR(TRITONSERVER_InferenceRequestSetResponseCallback(
+      request, nullptr, nullptr, nullptr, nullptr));
+  TRITONBACKEND_Request* backend_request =
+      reinterpret_cast<TRITONBACKEND_Request*>(request);
+  TRITONBACKEND_ResponseFactory* response_factory;
+  FAIL_TEST_IF_ERR(
+      TRITONBACKEND_ResponseFactoryNew(&response_factory, backend_request));
+
+  bool is_cancelled = true;
+  FAIL_TEST_IF_ERR(
+      TRITONSERVER_InferenceRequestIsCancelled(request, &is_cancelled));
+  ASSERT_FALSE(is_cancelled);
+
+  is_cancelled = true;
+  FAIL_TEST_IF_ERR(
+      TRITONBACKEND_RequestIsCancelled(backend_request, &is_cancelled));
+  ASSERT_FALSE(is_cancelled);
+
+  is_cancelled = true;
+  FAIL_TEST_IF_ERR(TRITONBACKEND_ResponseFactoryIsCancelled(
+      response_factory, &is_cancelled));
+  ASSERT_FALSE(is_cancelled);
+
+  is_cancelled = false;
+  FAIL_TEST_IF_ERR(TRITONSERVER_InferenceRequestCancel(request));
+
+  FAIL_TEST_IF_ERR(
+      TRITONBACKEND_RequestIsCancelled(backend_request, &is_cancelled));
+  ASSERT_TRUE(is_cancelled);
+
+  is_cancelled = false;
+  FAIL_TEST_IF_ERR(
+      TRITONSERVER_InferenceRequestIsCancelled(request, &is_cancelled));
+  ASSERT_TRUE(is_cancelled);
+
+  is_cancelled = false;
+  FAIL_TEST_IF_ERR(TRITONBACKEND_ResponseFactoryIsCancelled(
+      response_factory, &is_cancelled));
+  ASSERT_TRUE(is_cancelled);
+
+  FAIL_TEST_IF_ERR(TRITONBACKEND_ResponseFactoryDelete(response_factory));
```

Comment:
One test that might be nice here is to release the request from the backend - but the state of the response factory should still be cancelled / non cancelled -

pattern in some decoupled backends is to copy the request to an internal queue and release them while holding a reference to the response factory

Reply:
good point added.